File: consys-core/src/main/scala/de/tuda/stg/consys/core/store/TransactionContext.scala
package de.tuda.stg.consys.core.store

import scala.reflect.ClassTag

/**
 * Trait for transaction contexts to be used with a certain store type. Transaction
 * contexts store information about the current transaction.
 *
 * Transaction contexts are provided to developers when they start a transaction, and
 * are used for interacting with the replicated store.
 */
trait TransactionContext[StoreType <: Store] {
	val store : StoreType

	def replicate[T <: StoreType#ObjType : ClassTag](addr : StoreType#Addr, level : StoreType#Level, constructorArgs : Any*) : StoreType#RefType[T]
	def lookup[T <: StoreType#ObjType : ClassTag](addr : StoreType#Addr, level : StoreType#Level) : StoreType#RefType[T]


	/* Java interface for replicate */
	final def replicate[T <: StoreType#ObjType](addr : StoreType#Addr, level : StoreType#Level, clazz : Class[T], constructorArgs : Array[Any]) : StoreType#RefType[T] = {
		replicate(addr, level, constructorArgs : _*)(ClassTag(clazz))
	}
	/* Java interface for lookup */
	final def lookup[T <: StoreType#ObjType](addr : StoreType#Addr, level : StoreType#Level, clazz : Class[T]) : StoreType#RefType[T] = {
		lookup(addr, level)(ClassTag(clazz))
	}
}

File: consys-japi/src/main/java/de/tuda/stg/consys/japi/TransactionContext.java
package de.tuda.stg.consys.japi;

/**
 * Created on 27.01.20.
 *
 * @author Mirko Köhler
 */
public interface TransactionContext<Addr, Obj, Consistency> {

	<T extends Obj> Ref<T> replicate(Addr addr, Consistency level, Class<T> clazz, Object... constructorArgs);

	<T extends Obj> Ref<T> lookup(Addr addr, Consistency level, Class<T> clazz);

}
